Create Basic Data Types Tutorial

cuihtlauac opened the pull request:
Addresses predefined types, non mutable records, non mutually recursive and non generalized algebraic datatypes.

https://staging.ocaml.org/docs/basic-data-types

mseri commented on site/learn/tutorials/command-line_arguments.md:
```diff
@@ -2,51 +2,172 @@
+  automatically.
+
+* [Getopt](https://opam.ocaml.org/packages/getopt/) for OCaml is similar to
+  [GNU getopt](http://www.gnu.org/software/libc/manual/html_node/Getopt.html).
```
Maybe it is worth mentioning https://github.com/rbardou/clap/ and https://github.com/UnixJunkie/minicli

avsm commented:
Merged, thanks!

Commit:
Create Basic Data Types Tutorial

Addresses predefined types, non mutable records, non mutually recursive
and non generalized algebraic datatypes.

christinerose committed:
Initial Line Editing / grammar, syntax, formatting, etc.

christinerose reviewed (commented):
Some questions about clarity, minor syntax, and returned errors

cuihtlauac committed:
Update data/tutorials/language/0lg_08_basic_datatypes.md

Co-authored-by: Christine Rose <christinerose@users.noreply.github.com>

cuihtlauac committed:
Update data/tutorials/language/0lg_08_basic_datatypes.md

Co-authored-by: Christine Rose <christinerose@users.noreply.github.com>

Commit:
Address @christinerose feedback

christinerose reviewed (commented):
Got the rest to work now with putting the definitions again. I'll make a note about that.

cuihtlauac committed:
Update data/tutorials/language/0lg_08_basic_datatypes.md

Co-authored-by: Christine Rose <christinerose@users.noreply.github.com>

christinerose committed:
Update data/tutorials/language/0lg_08_basic_datatypes.md

Commit:
Misc typos

christinerose committed:
minor grammar

cuihtlauac committed:
Update data/tutorials/language/0lg_08_basic_datatypes.md

Co-authored-by: Christine Rose <christinerose@users.noreply.github.com>

asymmetric reviewed (commented):
I've left a review with some nits and some suggestions.

I am a total OCaml beginner, so take everything I say with a grain of salt, but I do have some experience in technical documentation.

christinerose committed:
Update data/tutorials/language/0lg_08_basic_datatypes.md

Co-authored-by: asymmetric <lorenzo@mailbox.org>

christinerose committed:
Update data/tutorials/language/0lg_08_basic_datatypes.md

Co-authored-by: asymmetric <lorenzo@mailbox.org>

christinerose committed:
Apply asymmetric's suggestions

Co-authored-by: asymmetric <lorenzo@mailbox.org>

christinerose committed:
Apply asymmetric's suggestions

Co-authored-by: asymmetric <lorenzo@mailbox.org>

christinerose committed:
Apply asymmetric's suggestions

Co-authored-by: asymmetric <lorenzo@mailbox.org>

christinerose committed:
Apply asymmetric's suggestions

Co-authored-by: asymmetric <lorenzo@mailbox.org>

christinerose committed:
Apply suggestions from code review

Co-authored-by: asymmetric <lorenzo@mailbox.org>

christinerose committed:
changes for consistency & clarity

christinerose committed:
added quote box for type annotation definition

Feel free to revise if it's not clear or I got something wrong.

Commit:
Suggestions from @asymmetric

christinerose committed:
remove repeated word

christinerose committed:
correct typo

Commit:
Review from @mobileink & &lindig

Commit:
Add missing link

christinerose committed:
minor grammar & typos

christinerose committed:
minor grammar

christinerose committed:
minor grammar

Commit:
Add missing word

christinerose committed:
clarify & tighten paragraph

Commit:
More polymorphism terminology

christinerose committed:
minor punctuation fix

christinerose committed:
slight rewording

asymmetric commented:
Is this ready to be merged? Just today I was looking for documentation on basic data types, and I'd love to see this at ocaml.org :)

sabine commented:
> Is this ready to be merged? Just today I was looking for documentation on basic data types, and I'd love to see this at ocaml.org :)

You can read it on staging.ocaml.org until it's merged.

Commit:
Import stuff from

Commit:
Add description

Commit:
Remove old datatypes & matching tutorial

Commit:
Change id

Commit:
Add epigram

Commit:
Team edits

christinerose committed:
minor grammar / syntax fixes

christinerose committed:
minor changes and UK spelling

christinerose commented on data/tutorials/language/0lg_00_basic_datatypes.md:
```diff
@@ -0,0 +1,844 @@
+Usually, `int` has 31 bits in 32-bit architectures and 63 in 64-bit architectures because one bit is reserved for OCaml's runtime operation. The standard library also provides [`Int32`](/api/Int32.html) and [`Int64`](/api/Int64.html) modules, which support platform independent operations on 32- and 64-bit signed integers. These modules are not detailed in this tutorial.
+
+There are no dedicated types for unsigned integers in OCaml. Bitwise operations on `int` treat the sign bit as the other bits. Binary operators use standard symbols. The signed remainder operator is written `mod`. There is no predefined power operator on integers in OCaml.
+
```
```suggestion

The `int` type is the default and basic type of integers in OCaml. When you enter a whole number, OCaml recognises it as an integer, as shown in this example:

```ocaml
# 42;;
- : int = 42
```

The `int` type represents platform-dependent signed integers. This means `int` does not always have same the number of bits. It depends on underlying platform characteristics, such as processor architecture or operating system. Operations on `int` values are provided by the [`Stdlib`](/api/Stdlib.html) and the [`Int`](/api/Int.html) modules.

Usually, `int` has 31 bits in 32-bit architectures and 63 in 64-bit architectures because one bit is reserved for OCaml's runtime operation. The standard library also provides [`Int32`](/api/Int32.html) and [`Int64`](/api/Int64.html) modules, which support platform-independent operations on 32- and 64-bit signed integers. However, these modules are not detailed in this tutorial.

OCaml doesn't have dedicated types for unsigned integers. Bitwise operations on `int` treat the sign bit as the other bits. Binary operators use standard symbols. The signed remainder operator is written `mod`. Predefined power operator on integers don't exist in OCaml.

```

"...treat the sign bit..."  should this be "signed bit" ?

cuihtlauac replied:
No it is “sign bit” like here: https://en.wikipedia.org/wiki/Sign_bit

christinerose commented on data/tutorials/language/0lg_00_basic_datatypes.md:
```diff
@@ -0,0 +1,844 @@
+
+#### Strings
+
+Strings are finite and fixed-sized sequences of values of type `char`. Strings are immutable, meaning it is impossible to change a character's value inside a string. The string concatenation operator symbol is `^`.
```
```suggestion
Strings are finite and fixed-sized sequences of `char` values. Strings are immutable, meaning it is impossible to change a character's value inside a string. The string concatenation operator symbol is `^`.
```

Would it be correct to say it this way? It gets rid of that extra prepositional phrase.

If so, could the same be done in li.139 & li.192?

cuihtlauac replied:
In a string, each value has type char, it is a sequence of char values.

christinerose replied:
Is that edited revision better?

christinerose commented on data/tutorials/language/0lg_00_basic_datatypes.md:
```diff
@@ -0,0 +1,844 @@
+
+#### Byte Sequences
+
+Byte sequences are finite and fixed-sized sequences of bytes. Each individual byte is represented by a `char` value. Byte sequences are mutable, meaning they can't be extended or shortened, but each component byte may be updated. Essentially, a byte sequence `bytes` is a mutable string that can't be printed. There is no way to write `bytes` literally, so they must be produced by a function.
```
```suggestion
Byte sequences are finite and fixed-sized. Each individual byte is represented by a `char` value. Byte sequences are mutable, meaning they can't be extended or shortened, but each component byte may be updated. Essentially, a byte sequence (type `bytes`) is a mutable string that can't be printed. There is no way to write `bytes` literally, so they must be produced by a function.
```

Would it work to remove "sequences of bytes" at the end of the first sentence, as it seems to repeat the beginning.

For this sentence: "a byte sequence `bytes` is a mutable string that can't be printed"
Is `bytes` a value type here? Or should it be "a byte sequence's `bytes`..."?

cuihtlauac replied:
It this doc, there are 4 kinds of sequences. All are finite.
- Strings (type `string`): elements are `char` and _can't_ be updated. Printable
- Byte Sequebce (type `bytes`): element are `char` and _can_ be updated. Non printable
- Array (type `'a array`): elements are `'a` and _can_ be updated. 
- List (type `'a list`): elements are `'a` and _can't_ be updated.

christinerose replied:
Is my amended suggestion better?

christinerose commented on data/tutorials/language/0lg_00_basic_datatypes.md:
```diff
@@ -0,0 +1,844 @@
+
+#### Results
+
+The `result` type can be used to express that the outcome of a function can be either success or failure. There are only two ways to build a result value: either using `Ok` or `Error` with the intended meaning. Both constructors can hold any kind of data. The `result` type is polymorphic, but it has two type parameters: one for `Ok` values and another for `Error` values.
```
```suggestion
The `result` type can be used to express that a function's outcome can be either success or failure. There are only two ways to build a result value: either using `Ok` or `Error` with the intended meaning. Both constructors can hold any kind of data. The `result` type is polymorphic, but it has two type parameters: one for `Ok` values and another for `Error` values.
```

christinerose reviewed (commented):
through li400

christinerose committed:
minor grammar

christinerose commented on data/tutorials/language/0lg_00_basic_datatypes.md:
````diff
@@ -0,0 +1,844 @@
+val commit_to_string' : commit -> string = <fun>
+```
+
+The `match … with …` construct needs to be passed an expression that is inspected. The `function …` is a special form of anonymous function taking a parameter and forwarding it to a `match … with …` construct as shown above.
````
```suggestion
We need to pass an inspected expression to the `match … with …` construct. The `function …` is a special form of anonymous function that takes a parameter and forwards it to a `match … with …` construct, as shown above.
```

Does this first sentence still say the same thing?

cuihtlauac replied:
Indeed, except in much better way.

christinerose commented on data/tutorials/language/0lg_00_basic_datatypes.md:
````diff
@@ -0,0 +1,844 @@
+type 'a list = [] | (::) of 'a * 'a list
+```
+
+The only bit of magic here is turning constructors into symbols. This is left unexplained in this tutorial. The types `bool` and `unit` also are regular variants, with the same magic:
````
```suggestion
The only bit of magic here is turning constructors into symbols, which we don't cover in this tutorial. The types `bool` and `unit` also are regular variants, with the same magic:
```

If not in this tutorial, where can we find it?

cuihtlauac replied:
Need to add that into the “operators” tutorial

christinerose commented on data/tutorials/language/0lg_00_basic_datatypes.md:
````diff
@@ -0,0 +1,844 @@
+val map : ('a -> 'b) -> 'a tree -> 'b tree = <fun>
+```
+
+In the OCaml community as well as in the larger functional programming community, the word *polymorphism* is used loosely. It is applied to things working in a similar fashion with various types. Several features of OCaml are polymorphic, in the broad sense, each is using a particular form of it, including its own name. In summary, OCaml has several forms of polymorphism. In most cases, the disti ...
````
```suggestion
In the OCaml community as well as in the larger functional programming community, the word *polymorphism* is used loosely. It is applied to things working in a similar fashion with various types. Several features of OCaml are polymorphic, in the broad sense. Each uses a particular form of it, including its own name. In summary, OCaml has several forms of polymorphism. In most cases, the distinction between those concepts is blurred, but it sometimes needs to distinguish them. Here are the terms applicable to datatypes:
```

When you say "each is using a particular form of it..." -- each what? Each feature of OCaml?

cuihtlauac replied:
_Features of ocaml which are polymorphic_

christinerose replied:
I rewrote it as: "In the OCaml community, as well as in the larger functional programming community, the word *polymorphism* is used loosely. It is applied to things working in a similar fashion with various types. Several features of OCaml are polymorphic in this broad sense. Each polymorphic feature uses a particular form of polymorphism, including its own name. In summary, OCaml has several forms of polymorphism. In most cases, the distinction between those concepts is blurred, but it sometimes necessary to distinguish them. Here are the terms applicable to datatypes:"

christinerose commented on data/tutorials/language/0lg_00_basic_datatypes.md:
```diff
@@ -0,0 +1,844 @@
+
+### Records
+
+Records are like tuples in that several values are bundled together. In a tuple, elements are identified by their position in the corresponding product type. They are either first, second, third, or at some other position. In a record, each element has a name, it's a field. That's why record types must be declared before being used.
```
Let's clarify what this sentence means: "In a record, each element has a name, it's a field."

Each element has a name, and that name is a field? The element is a field? There is a field for the name?

cuihtlauac replied:
A field is a pair, a name and a value. The fields (i.e. the name-value pairs) are the elements composing a record.

christinerose replied:
How about this:  "In a record, each element has a *field* that contains a name and a value."

Does this make it clear that the name-value pair is known as a field?

Or should it be: "In a record, each element has a name and a value. This name-value pair is known as a field."

christinerose replied:
(I changed it to the first sentence, but we can further revise it if necessary.)

cuihtlauac replied:
> In a record, each element has a name and a value. This name-value pair is known as a field.

This is the best wording

christinerose commented on data/tutorials/language/0lg_00_basic_datatypes.md:
```diff
@@ -0,0 +1,844 @@
+
+1. Records **should not** be defined using this technique. It is only demonstrated here to further illustrate the expressive strength of OCaml variants.
+
+1. This way, to define records **may** be applied to _Generalised Algebraic Data Types_, which is the subject of another tutorial.
```
To define record may be applied to...
Defining records may be applied to...
The definition of records may be applied to...

Which of these is correct?

cuihtlauac replied:
It is the coma which is misleading and should go away. This is like: "This way to do the hula hoop ..."

christinerose replied:
So like: "Defining records in this way may be applied to ...."
"This method of defining records may be applied to..."

In English we'd say something more like: "This is the way to do the hula hoop" or "The way to do the hula hoop is to rotate your hips...."

cuihtlauac replied:
I understand, thanks for the clarification. I like 
> Defining records in this way may be applied to

BTW, this has moved to the `0lg_11_records_as_variants.md` file

christinerose commented on data/tutorials/language/0lg_00_basic_datatypes.md:
```diff
@@ -0,0 +1,844 @@
+
+## Conclusion
+
+This tutorial has provided a comprehensive overview of the basic data types in OCaml and their usage. We have explored the built-in types, such as integers, floats, characters, lists, tuples, and strings, and the user-defined types: records and variants. Records and tuples are mechanisms for grouping heterogeneous data into cohesive units. Variants are a mechanism for exposing heterogeneous data  ...
```
Should the comments/definitions of Records and tuples be in their respective sections (both in this paragraph and the one beneath)? Is this repetitive here, or is the intention to summarise what was already said to remind the reader? If so, I'd like to reword it slightly. 

Currently it reads as if we're getting this information for the first time.

cuihtlauac replied:
I need to rewrite the whole conclusion

christinerose reviewed (commented):
400 - end

christinerose committed:
minor grammar / syntax fixes

asymmetric commented on data/tutorials/language/0lg_00_basic_datatypes.md:
```diff
@@ -23,26 +16,28 @@ OCaml is a statically and strongly typed programming language. It is also an exp
 
-OCaml provides several other types, but they all are extensions of those presented in this tutorial. Types that are in the scope of this tutorial are all the basic constructors and the most common predefined types.
+The OCaml type system aggregates several type systems, also known as disciplines:
+- A [nominal type system](https://en.wikipedia.org/wiki/Nominal_type_system) is used for predefined types, variants, and functions. Historically, it the first system, directly coming from the type system used the [ML](https://en.wikipedia.org/wiki/ML_(programming_language)) programming language, which is OCaml accestor. By default, this what is meant by OCaml type system, and it is also the scope ...
```
```suggestion
- A [nominal type system](https://en.wikipedia.org/wiki/Nominal_type_system) is used for predefined types, variants, and functions. Historically, it the first system, directly coming from the type system used the [ML](https://en.wikipedia.org/wiki/ML_(programming_language)) programming language, which is OCaml's ancestor. By default, this is what is meant by OCaml type system, and it is also the scope of this tutorial.
```

cuihtlauac replied:
Manually merged

asymmetric commented on data/tutorials/language/0lg_00_basic_datatypes.md:
```diff
@@ -23,26 +16,28 @@ OCaml is a statically and strongly typed programming language. It is also an exp
-  * Another for objects and classes
-
-Each discipline is using a different type-checking algorithm.
+The goal of this tutorial will provide capabilities to handle data from predefined, variant and record types. This importantly includes pattern matching on those types.
```
```suggestion
The goal of this tutorial is to teach skills on how to handle data from predefined, variant and record types. This importantly includes pattern matching on those types.
```

cuihtlauac replied:
Manually merged and moved into the first paragraph.

asymmetric commented on data/tutorials/language/0lg_00_basic_datatypes.md:
```diff
@@ -23,26 +16,28 @@ OCaml is a statically and strongly typed programming language. It is also an exp
+  * One for polymorphic variants
+  * Another for objects and classes
+
+OCaml also provides Generalized Algebraic Data Types, which are an extensions of those presented in this tutorial. Types that are in the scope of this tutorial are the most basic  ones.
```
```suggestion
OCaml also provides [Generalized Algebraic Data Types](https://en.wikipedia.org/wiki/Generalized_algebraic_data_type), which are an extensions of those presented in this tutorial. Types that are in the scope of this tutorial are the most basic  ones.
```

cuihtlauac replied:
Manually merged link somewhere else.

Commit:
Merging team edits

Commit:
Manually merge of comments for @asymmetric

Commit:
Work introduction and conclusion

Commit:
Cast away records as variants

christinerose committed:
Slight rewrites based on Cuit's answers

Commit:
Add conclusion

christinerose commented on data/tutorials/language/0lg_00_basic_datatypes.md:
```diff
@@ -0,0 +1,809 @@
+
+### Records
+
+Records are like tuples in that several values are bundled together. In a tuple, elements are identified by their position in the corresponding product type. They are either first, second, third, or at some other position. In a record, each element has a *field* that contains a name and a value. That's why record types must be declared before being used.
```
```suggestion
Records are like tuples in that several values are bundled together. In a tuple, elements are identified by their position in the corresponding product type. They are either first, second, third, or at some other position. In a record, each element has a name and a value. This name-value pair is known as a field. That's why record types must be declared before being used.
```

christinerose committed:
Clarify wording for records

Commit:
Turn Record as Variants into another tutorial

Candidate title: Advanced Topics on Variants and Products

christinerose committed:
line editing through li. 60

christinerose commented on data/tutorials/language/0lg_00_basic_datatypes.md:
```diff
@@ -0,0 +1,807 @@
+
+The `int` type represents platform-dependent signed integers. This means `int` does not always have same the number of bits. It depends on underlying platform characteristics, such as processor architecture or operating system. Operations on `int` values are provided by the [`Stdlib`](/api/Stdlib.html) and the [`Int`](/api/Int.html) modules.
+
+Usually, `int` has 31 bits in 32-bit architectures and 63 in 64-bit architectures because one bit is reserved for OCaml's runtime operation. The standard library also provides [`Int32`](/api/Int32.html) and [`Int64`](/api/Int64.html) modules, which support platform independent operations on 32- and 64-bit signed integers. These modules are not detailed in this tutorial.
```
Can part (or all) of the paragraph be deleted since it talks about things not covered in this tutorial?

christinerose committed:
Syntax tightening

christinerose committed:
Update data/tutorials/language/0lg_00_basic_datatypes.md

christinerose committed:
Update data/tutorials/language/0lg_00_basic_datatypes.md

christinerose committed:
tightened syntax, removed repetition

Commit:
Add reference to tagged union

christinerose committed:
change alpha definition to reminder

christinerose committed:
changed examples a, b, c to x, y, z

in order to avoid confusion with "a" and `'a`

cuihtlauac commented on data/tutorials/language/0lg_00_basic_datatypes.md:
````diff
@@ -0,0 +1,808 @@
+- : int = 2
+```
+
+In the above expressions, `[1; 2; 3]` is the value that is matched over. Each expression between the `|` and `->` symbols is a pattern. They are expressions of type list, only formed using `[]`, `::`, and variables names that represent various shapes a list may have. When the pattern is `[]`, it means “if the list is empty.” The pattern `x :: u` means “if the list contains data, let `x` be the fi ...
````
@christinerose 
```suggestion
In the above expressions, `[1; 2; 3]` is the value that is matched over. Each expression between the `|` and `->` symbols is a pattern. They are expressions of type list, only formed using `[]`, `::`, and variables names that represent various shapes a list may have. The pattern `[]` means “if the list is empty.” The pattern `x :: u` means “if the list contains data, let `x` be the first element of the list and `u` be the rest of the list.” Expressions at the right of the `->` symbol are the results returned in each corresponding case.
```

Should both sentences on both kind of patterns have the same structure? Both either with "when" or without?

christinerose replied:
Nice! Yes!

cuihtlauac commented on data/tutorials/language/0lg_00_basic_datatypes.md:
````diff
@@ -0,0 +1,808 @@
+- : int * char = (3, 'a')
+```
+
+That pair contains the integer `3` and the character `'a'`; its type is `int * char`. The `*` symbol stands for _product type_.
````
````suggestion
# (3, 'K');;
- : int * char = (3, 'K')
```

That pair contains the integer `3` and the character `'K'`; its type is `int * char`. The `*` symbol stands for _product type_.
````

Removing more 'a'

cuihtlauac commented on data/tutorials/language/0lg_00_basic_datatypes.md:
````diff
@@ -0,0 +1,808 @@
+val f : 'a * 'b * 'c * 'd -> 'c = <fun>
+```
+
+Note that types `int * char * bool`, `int * (char * bool)`, and `(int * char) * bool` are not the same. The values `(42, 'a', true)`, `(42, ('a', true))`, and `((42, 'a'), true)` are not equal. In mathematical language, the product type operator `*` is not associative.
````
```suggestion
Note that types `int * char * bool`, `int * (char * bool)`, and `(int * char) * bool` are not the same. The values `(42, 'a', true)`, `(42, ('a', true))`, and `((42, 'a'), true)` are not equal. In mathematical language, the product type operator `*` is not _associative_.
```

This use of this term (I believe)

cuihtlauac commented on data/tutorials/language/0lg_00_basic_datatypes.md:
````diff
@@ -0,0 +1,808 @@
+- : string = "This is really disco!"
+```
+
+The first expression is another anonymous function. It is the identity function, it can be applied to anything, and it returns its argument unchanged. This means that its parameter can be of any type, and its result has the same type. The same code can be applied to data of different types. This is called _polymorphism_.
````
```suggestion
The first expression is another anonymous function. It is the _identity_ function, it can be applied to anything, and it returns its argument unchanged. This means that its parameter can be of any type, and its result has the same type. The same code can be applied to data of different types. This is called _polymorphism_.
```

This is the first use of the math term “identity”

cuihtlauac commented on data/tutorials/language/0lg_00_basic_datatypes.md:
```diff
@@ -0,0 +1,808 @@
+
+The first expression is another anonymous function. It is the identity function, it can be applied to anything, and it returns its argument unchanged. This means that its parameter can be of any type, and its result has the same type. The same code can be applied to data of different types. This is called _polymorphism_.
+
+Remember, the `'a` is a _type variable_, so values of any type can be passed to the function and replaces the type variable. Its identity has the same input and output type, whatever it may be.
```
```suggestion
Remember, the `'a` is a _type variable_, so values of any type can be passed to the function and replaces the type variable. The identity function has the same input and output type, whatever it may be.
```

cuihtlauac commented on data/tutorials/language/0lg_00_basic_datatypes.md:
```diff
@@ -0,0 +1,808 @@
+
+`(int, bool) pair` would be written `int * bool`, and `Pair (42, true)` would be written `(42, true)`. From the developer's perspective, everything happens as if such a type were declared for every possible product shape. This is what allows pattern matching on products.
+
+Even integers and floats can be seen as enumerated-like variant types, with many constructors and funky syntactic sugar, which allows pattern matching on those types.
```
Nice, you didn't killed “funky”.

Is this the right place for an easter egg?
Maybe this one: https://youtu.be/O0_H3F84Yjk

christinerose replied:
HELL YES! 😂 🕺

christinerose committed:
Update data/tutorials/language/0lg_00_basic_datatypes.md

Co-authored-by: Cuihtlauac Alvarado <cuihtlauac@users.noreply.github.com>

christinerose committed:
Update data/tutorials/language/0lg_00_basic_datatypes.md

Co-authored-by: Cuihtlauac Alvarado <cuihtlauac@users.noreply.github.com>

christinerose committed:
Update data/tutorials/language/0lg_00_basic_datatypes.md

Co-authored-by: Cuihtlauac Alvarado <cuihtlauac@users.noreply.github.com>

christinerose committed:
Update data/tutorials/language/0lg_00_basic_datatypes.md

Co-authored-by: Cuihtlauac Alvarado <cuihtlauac@users.noreply.github.com>

christinerose committed:
Update data/tutorials/language/0lg_00_basic_datatypes.md

Co-authored-by: Cuihtlauac Alvarado <cuihtlauac@users.noreply.github.com>

christinerose committed:
added fun easter egg video

christinerose committed:
minor line editing

christinerose commented on data/tutorials/language/0lg_00_basic_datatypes.md:
```diff
@@ -0,0 +1,808 @@
+# let letter = [| 'v'; 'x'; 'y'; 'z' |];;
+val letter : char array = [|'v'; 'x'; 'y'; 'z'|]
+
+# letter.(2) <- '3';;
```
Since the arrow is pointing the other way for the first time (that I remember), should we speak to it? After some pondering, I saw what it was doing (passing `3` to the 2nd position of `letter`, but it confused me at first.

cuihtlauac replied:
I'm writing something on this.

christinerose commented on data/tutorials/language/0lg_00_basic_datatypes.md:
````diff
@@ -0,0 +1,808 @@
+- : char array = [|'v'; 'x'; '3'; 'z'|]
+```
+
+Operations on arrays are provided by the [`Array`](/api/Array.html) module. There is a dedicated tutorial on Arrays.
````
```suggestion
Operations on arrays are provided by the [`Array`](/api/Array.html) module. There is a dedicated [tutorial on Arrays](https://ocaml.org/docs/arrays).
```

christinerose commented on data/tutorials/language/0lg_00_basic_datatypes.md:
```diff
@@ -0,0 +1,808 @@
+
+Like arrays, lists are finite sequences of values of the same type. They are polymorphic, too. However, lists are extensible, immutable, and don't support direct access to all the values they contain. Lists play a central role in functional programming, so they have a [dedicated tutorial](/docs/lists).
+
+Operations on lists are provided by the [`List`](/api/List.html) module. The `List.append` function concatenates two lists. It can be used as an operator with the symbol `@`.
```
Do you want to tell them how to pronounce `@`?

cuihtlauac replied:
Damm! I don't how to pronounce that!

christinerose commented on data/tutorials/language/0lg_00_basic_datatypes.md:
```diff
@@ -0,0 +1,808 @@
+
+Like the product type symbol `*`, the function type symbol `->` is not associative. The following two types are not the same:
+- `(int -> int) -> int` : This function takes a function of type `int -> int` as a parameter and returns an `int` as the result.
+- `int -> (int -> int)` : This function takes an `int` asa  parameter and returns a function of type `int -> int` as the result.
```
Is this similar to what we covered in Functions & Variables? About how `->` associates to the right? (li 562)

cuihtlauac replied:
Yes it is.

christinerose commented on data/tutorials/language/0lg_00_basic_datatypes.md:
````diff
@@ -0,0 +1,808 @@
+- : string -> unit = <fun>
+```
+
+The function `read_line` reads an end-of-line terminated sequence of characters from standard input and returns it as a string. Reading input begins when `()` is passed.
````
Should we show an example here? Perhaps what it looks like to pass `()` to `read_line`?

cuihtlauac replied:
Yes, but there is a hicup. It would be nice to have something like this:
```ocaml
 # read_line ();;
foo bar
- : string = "foo bar"
```

However, we have to say “replace `foo bar` by your own text, followed by end of line”

christinerose commented on data/tutorials/language/0lg_00_basic_datatypes.md:
```diff
@@ -0,0 +1,808 @@
+
+### Variants
+
+Variants are also called [_tagged unions_](https://en.wikipedia.org/wiki/Tagged_union). They relate to the concept of [_disjoint union_](https://en.wikipedia.org/wiki/Disjoint_union).
```
Do you want an example here like you have for Enumerated, etc.?

cuihtlauac replied:
This is an introduction for the rest of this section. Examples will come later.

christinerose reviewed (commented):
Looks to me like it's pretty much ready to merge. Just a few questions/comments. Once those are addressed or dismissed, I think we can merge with Values & Functions! 🎉

Commit:
Edits

Commit:
Edits, cont'd

Commit:
File renaming

christinerose committed:
add example

christinerose commented on data/tutorials/language/0lg_01_basic_datatypes.md:
````diff
@@ -0,0 +1,814 @@
+- : char array = [|'v'; 'x'; 'A'; 'z'|]
+```
+
+The left-arrow `<-` is the array update operator. Above, it means the cell at index 2 is set to value `'A'`. It is the same as writing `Array.set letter 2 'A'`. Array update is a side effect, the unit value is returned.
````
```suggestion
The left-arrow `<-` is the array update operator. Above, it means the cell at index 2 is set to value `'A'`. It is the same as writing `Array.set letter 2 'A'`. Array update is a side effect, and the unit value is returned.
```

christinerose committed:
slight correction

Commit:
Minor edits

Commit:
Various fixes

Commit:
More edits

Commit:
Records are variants

Commit:
Typos

Commit:
Typos, cont'd

Commit:
Break long line
